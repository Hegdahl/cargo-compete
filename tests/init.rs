use cargo_compete::{shell::Shell, Opt};
use duct::cmd;
use ignore::WalkBuilder;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;
use std::{path::Path, str};
use structopt::StructOpt as _;

#[test]
fn no_crate() -> anyhow::Result<()> {
    let (output, tree) = run("\n1\n")?;
    assert_snapshot!("no_crate_output", output);
    assert_json_snapshot!("no_crate_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[test]
fn use_crate() -> anyhow::Result<()> {
    let (output, tree) = run("\n2\n")?;
    assert_snapshot!("use_crate_output", output);
    assert_json_snapshot!("use_crate_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[test]
fn use_crate_via_bianry() -> anyhow::Result<()> {
    let (output, tree) = run("\n3\n")?;
    assert_snapshot!("use_crate_via_bianry_output", output);
    assert_json_snapshot!("use_crate_via_bianry_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(input: &'static str) -> anyhow::Result<(String, serde_json::Value)> {
    let workspace = tempfile::Builder::new()
        .prefix("cargo-compete-test-workspace")
        .tempdir()?;

    let (output_file, output) = tempfile::Builder::new()
        .prefix("cargo-compete-test-output")
        .tempfile()?
        .into_parts();

    println!("{}", cmd!("git", "init", workspace.path()).read()?);

    let Opt::Compete(opt) = Opt::from_iter_safe(&["", "compete", "i"])?;

    cargo_compete::run(
        opt,
        cargo_compete::Context {
            cwd: workspace.path().to_owned(),
            shell: &mut Shell::from_read_write(Box::new(input.as_bytes()), Box::new(output_file)),
        },
    )?;

    let output_masked = std::fs::read_to_string(&output)?
        .replace(workspace.path().to_str().unwrap(), "{{ cwd }}")
        .replace(std::path::MAIN_SEPARATOR, "{{ separator }}");

    let tree = tree(workspace.as_ref())?;

    workspace.close()?;
    output.close()?;

    Ok((output_masked, tree))
}

fn tree(path: &Path) -> anyhow::Result<serde_json::Value> {
    let mut tree = serde_json::Map::new();

    for entry in WalkBuilder::new(path)
        .git_ignore(false)
        .sort_by_file_name(Ord::cmp)
        .build()
    {
        let entry = entry?;

        let components = entry
            .path()
            .strip_prefix(path)?
            .iter()
            .map(|p| p.to_str().unwrap())
            .collect::<Vec<_>>();

        let mut tree = &mut tree;
        if entry.path().is_dir() {
            for component in components {
                tree = tree
                    .entry(component)
                    .or_insert_with(|| json!({}))
                    .as_object_mut()
                    .unwrap();
            }
        } else if let [components @ .., file_name] = &*components {
            for &component in components {
                tree = tree
                    .entry(component)
                    .or_insert_with(|| json!({}))
                    .as_object_mut()
                    .unwrap();
            }
            tree.insert(
                (*file_name).to_owned(),
                json!(std::fs::read_to_string(entry.path())?),
            );
        } else {
            panic!();
        }
    }

    Ok(serde_json::Value::Object(tree))
}
