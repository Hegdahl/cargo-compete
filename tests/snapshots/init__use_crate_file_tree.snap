---
source: tests/init.rs
expression: tree
---
{
  "atcoder": {
    "Cargo.lock": "..",
    "Cargo.toml": "[workspace]\nmembers = [\"cargo-compete-template\"]\nexclude = []\n",
    "cargo-compete-template": {
      "Cargo.toml": "[package]\nname = \"cargo-compete-template\"\nversion = \"0.1.0\"\nedition = \"2018\"\npublish = false\n\n[[bin]]\nname = \"cargo-compete-template\"\npath = \"src/main.rs\"\n\n[dependencies]\nnum = \"=0.2.1\"\nnum-bigint = \"=0.2.6\"\nnum-complex = \"=0.2.4\"\nnum-integer = \"=0.1.42\"\nnum-iter = \"=0.1.40\"\nnum-rational = \"=0.2.4\"\nnum-traits = \"=0.2.11\"\nnum-derive = \"=0.3.0\"\nndarray = \"=0.13.0\"\nnalgebra = \"=0.20.0\"\nalga = \"=0.9.3\"\nlibm = \"=0.2.1\"\nrand = { version = \"=0.7.3\", features = [\"small_rng\"] }\ngetrandom = \"=0.1.14\"\nrand_chacha = \"=0.2.2\"\nrand_core = \"=0.5.1\"\nrand_hc = \"=0.2.0\"\nrand_pcg = \"=0.2.1\"\nrand_distr = \"=0.2.2\"\npetgraph = \"=0.5.0\"\nindexmap = \"=1.3.2\"\nregex = \"=1.3.6\"\nlazy_static = \"=1.4.0\"\nordered-float = \"=1.0.2\"\nascii = \"=1.0.0\"\npermutohedron = \"=0.2.4\"\nsuperslice = \"=1.0.0\"\nitertools = \"=0.9.0\"\nitertools-num = \"=0.1.3\"\nmaplit = \"=1.0.2\"\neither = \"=1.5.3\"\nim-rc = \"=14.3.0\"\nfixedbitset = \"=0.2.0\"\nbitset-fixed = \"=0.1.0\"\nproconio = { version = \"=0.3.6\", features = [\"derive\"] }\ntext_io = \"=0.1.8\"\nwhiteread = \"=0.5.0\"\nrustc-hash = \"=1.1.0\"\nsmallvec = \"=1.2.0\"\n",
      "src": {
        "main.rs": "use proconio::{fastout, input};\n\n#[fastout]\nfn main() {\n    input! {\n        n: usize,\n    }\n}\n"
      }
    },
    "compete.toml": "# How to manage new workspace members (\"include\" | \"exclude\" | \"focus\")\n#\n# - `skip`:    Does not modify `[workspace]`\n# - `include`: Adds the package to `workspace.members`\n# - `exclude`: Adds the package to `workspace.exclude` and create a symlink to the `compete.toml`\n# - `focus`:   Adds the package to `workspace.members` and remove the other from both of `workspace.{members, exclude}`\nnew-workspace-member = \"include\"\n\n# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `problem`:      Problem index (e.g. \"a\", \"b\")\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"./testcases/{{ contest }}/{{ problem | kebabcase }}.yml\"\n#test-suite = \"{{ manifest_dir }}/testcases/{{ problem | kebabcase }}.yml\"\n\n# Open files with the command (`jq` command)\n#\n# VSCode:\n#open = '[\"code\"] + (.paths | map([.src, .test_suite]) | flatten) + [\"-a\", .manifest_dir]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[template]\nplatform = \"atcoder\"\nmanifest = \"./cargo-compete-template/Cargo.toml\"\nsrc = \"./cargo-compete-template/src/main.rs\"\n\n#[submit-via-binary]\n#target = \"x86_64-unknown-linux-musl\"\n##cross = \"cross\"\n#strip = \"strip\"\n##upx = \"upx\"\n",
    "rust-toolchain": "1.42.0\n"
  },
  "codeforces": {
    "Cargo.lock": "..",
    "Cargo.toml": "[workspace]\nmembers = [\"cargo-compete-template\"]\nexclude = []\n",
    "cargo-compete-template": {
      "Cargo.toml": "[package]\nname = \"cargo-compete-template\"\nversion = \"0.1.0\"\nedition = \"2018\"\npublish = false\n\n[[bin]]\nname = \"cargo-compete-template\"\npath = \"src/main.rs\"\n",
      "src": {
        "main.rs": "use std::io::{self, Read as _};\n\nfn main() {\n    let mut input = \"\".to_owned();\n    io::stdin().read_to_string(&mut input).unwrap();\n    let mut input = input.split_whitespace();\n    macro_rules! read(\n        ([$tt:tt]) => (read!([$tt; read!(usize)]));\n        ([$tt:tt; $n:expr]) => ((0..$n).map(|_| read!($tt)).collect::<Vec<_>>());\n        (($($tt:tt),+)) => (($(read!($tt)),*));\n        (Usize1) => (read!(usize) - 1);\n        (Byte) => (read!(char) as u8);\n        (Bytes) => (read!(String).into_bytes());\n        ($ty:ty) => (input.next().unwrap().parse::<$ty>().unwrap());\n    );\n\n    let n = read!(usize);\n}\n"
      }
    },
    "compete.toml": "# How to manage new workspace members (\"include\" | \"exclude\" | \"focus\")\n#\n# - `skip`:    Does not modify `[workspace]`\n# - `include`: Adds the package to `workspace.members`\n# - `exclude`: Adds the package to `workspace.exclude` and create a symlink to the `compete.toml`\n# - `focus`:   Adds the package to `workspace.members` and remove the other from both of `workspace.{members, exclude}`\nnew-workspace-member = \"include\"\n\n# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `problem`:      Problem index (e.g. \"a\", \"b\")\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"./testcases/{{ contest }}/{{ problem | kebabcase }}.yml\"\n#test-suite = \"{{ manifest_dir }}/testcases/{{ problem | kebabcase }}.yml\"\n\n# Open files with the command (`jq` command)\n#\n# VSCode:\n#open = '[\"code\"] + (.paths | map([.src, .test_suite]) | flatten) + [\"-a\", .manifest_dir]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[template]\nplatform = \"codeforces\"\nmanifest = \"./cargo-compete-template/Cargo.toml\"\nsrc = \"./cargo-compete-template/src/main.rs\"\n\n#[submit-via-binary]\n#target = \"x86_64-unknown-linux-musl\"\n##cross = \"cross\"\n#strip = \"strip\"\n##upx = \"upx\"\n",
    "rust-toolchain": "1.42.0\n"
  },
  "yukicoder": {
    "Cargo.lock": "..",
    "Cargo.toml": "[workspace]\nmembers = [\"cargo-compete-template\"]\nexclude = []\n",
    "cargo-compete-template": {
      "Cargo.toml": "[package]\nname = \"cargo-compete-template\"\nversion = \"0.1.0\"\nedition = \"2018\"\npublish = false\n\n[[bin]]\nname = \"cargo-compete-template\"\npath = \"src/main.rs\"\n",
      "src": {
        "main.rs": "use std::io::{self, Read as _};\n\nfn main() {\n    let mut input = \"\".to_owned();\n    io::stdin().read_to_string(&mut input).unwrap();\n    let mut input = input.split_whitespace();\n    macro_rules! read(\n        ([$tt:tt]) => (read!([$tt; read!(usize)]));\n        ([$tt:tt; $n:expr]) => ((0..$n).map(|_| read!($tt)).collect::<Vec<_>>());\n        (($($tt:tt),+)) => (($(read!($tt)),*));\n        (Usize1) => (read!(usize) - 1);\n        (Byte) => (read!(char) as u8);\n        (Bytes) => (read!(String).into_bytes());\n        ($ty:ty) => (input.next().unwrap().parse::<$ty>().unwrap());\n    );\n\n    let n = read!(usize);\n}\n"
      }
    },
    "compete.toml": "# How to manage new workspace members (\"include\" | \"exclude\" | \"focus\")\n#\n# - `skip`:    Does not modify `[workspace]`\n# - `include`: Adds the package to `workspace.members`\n# - `exclude`: Adds the package to `workspace.exclude` and create a symlink to the `compete.toml`\n# - `focus`:   Adds the package to `workspace.members` and remove the other from both of `workspace.{members, exclude}`\nnew-workspace-member = \"include\"\n\n# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `problem`:      Problem index (e.g. \"a\", \"b\")\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"./testcases/{{ contest }}/{{ problem | kebabcase }}.yml\"\n#test-suite = \"{{ manifest_dir }}/testcases/{{ problem | kebabcase }}.yml\"\n\n# Open files with the command (`jq` command)\n#\n# VSCode:\n#open = '[\"code\"] + (.paths | map([.src, .test_suite]) | flatten) + [\"-a\", .manifest_dir]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[template]\nplatform = \"yukicoder\"\nmanifest = \"./cargo-compete-template/Cargo.toml\"\nsrc = \"./cargo-compete-template/src/main.rs\"\n\n#[submit-via-binary]\n#target = \"x86_64-unknown-linux-musl\"\n##cross = \"cross\"\n#strip = \"strip\"\n##upx = \"upx\"\n",
    "rust-toolchain": "1.44.1\n"
  }
}
